_posts/2022-01-03-logical-ops-on-java-collections.md

```diff
@@ -0,0 +1,344 @@
+        
+        if(!intersectElements.isEmpty()) {
+            return intersectElements;
+        }else {
+            return Collections.emptyList();
+        }
+        
+    }
+}
+```
+For finding the intersection of two collections, we run the `filter()` method on the first collection to identify and collect the matching elements from the second collection. 
+
+The corresponding unit test looks like this:
+
+```java
+class CollectionHelperTest {
+    
+    private CollectionHelper collectionHelper;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        collectionHelper = new CollectionHelper();
+    }
+    
+    @Test
+    void testIntersection() {
+        List<Integer> intersection = collectionHelper.intersection(
+                List.of(9,8,5,4,7, 15, 15), 
+                List.of(1,3,99,4,7));
+        
+        Assertions.assertArrayEquals(
+                List.of(4,7).toArray(), 
+                intersection.toArray());
+    }
+}
+```
+
+## Subtracting one Collection from Another
+In this example we are again using Java's `Stream` class to find the collection of different elements contained in two collections:
+
+![subtracting two collections](/assets/img/posts/logical-ops-java-coll/minus.png)
+
+```java
+public class CollectionHelper {
+    
+    public List<Integer> subtract(
+        final List<Integer> collA, 
+        final List<Integer> collB){
+
+        List<Integer> intersectElements = intersection(collA,collB);
+        
+        List<Integer> subtractedElements = collA.stream()
+            .filter(element->!intersectElements
+                .contains(element))
+            .collect(Collectors.toList());
+        
+        if(!subtractedElements.isEmpty()) {
+            return subtractedElements;
+        }else {
+            return Collections.emptyList();
+        }
+        
+    }
+
+    public List<Integer> intersection(
+                            final List<Integer> collA, 
+                            final List<Integer> collB){
+        // see implementation above
+    }
+}
+```
+
+The difference between to collections is found in two steps:
+1. Finding the common set of elements with the `intersection()` method seen above.
+2. Applying the `filter()` method of the `Stream` class on the first collection to exclude those elements.
+
+The corresponding unit test looks like this:
+
+```java
+class CollectionHelperTest {
+    
+    CollectionHelper collectionHelper;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        collectionHelper = new CollectionHelper();
+    }
+
+ 
+    @Test
+    void testSubtraction() {
+        List<Integer> sub = collectionHelper.subtract(
+                List.of(9,8,5,4,7, 15, 15), 
+                List.of(1,3,99,4,7));
+        
+        
+        Assertions.assertArrayEquals(
+                List.of(9,8,5,15,15).toArray(), 
+                sub.toArray());
+    }
+
+}
+```
+
+## Extract Subset - Split a List into Two Sublists
```

Comment:
This isn't really a "logical operation" on a list, so it doesn't match to the title of the article. So, either we rename the article to "Common Operations on Collections" (and add other common operations like creating a list from a comma-separated string, joining items from a list with a separator to create a string, and others that you can think of) or remove this section.